instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/event_handler.rb

```diff
@@ -0,0 +1,277 @@
+        #           GLOBAL_LOGGER.info("otel.trace_id=#{rack_span.context.hex_trace_id} otel.span_id=#{rack_span.context.hex_span_id}")
+        #         end
+        #         [200, { 'Content-Type' => 'text/plain' }, body]
+        #       end
+        #     }
+        #   end
+        #
+        # @see Rack::Events
+        # @see OpenTelemetry::Instrumentation::Rack.current_span
+        class EventHandler
+          include ::Rack::Events::Abstract
+
+          TOKENS_KEY = 'otel.context.tokens'
+          GOOD_HTTP_STATUSES = (100..499).freeze
+
+          # Creates a server span for this current request using the incoming parent context
+          # and registers them as the {current_span}
+          #
+          # @param [Rack::Request] The current HTTP request
+          # @param [Rack::Response] This is nil in practice
+          # @return [void]
+          def on_start(request, _)
+            return if untraced_request?(request.env)
+
+            extracted_context = extract_remote_context(request)
+            if record_frontend_span?
+              request_start_time = OpenTelemetry::Instrumentation::Rack::Util::QueueTime.get_request_start(request.env)
+              unless request_start_time.nil?
+                frontend_span = tracer.start_span(
+                  'http_server.proxy',
+                  with_parent: extracted_context,
+                  start_timestamp: request_start_time,
+                  kind: :server
+                )
+
+                frontend_context = OpenTelemetry::Trace.context_with_span(frontend_span, parent_context: extracted_context)
+              end
+            end
+            parent_context = frontend_context || extracted_context
+
+            span = tracer.start_span(
+              create_request_span_name(request),
+              with_parent: parent_context,
+              kind: frontend_context.nil? ? :server : :internal,
+              attributes: request_span_attributes(request.env)
+            )
+            ctx = OpenTelemetry::Trace.context_with_span(span)
+            rack_ctx = OpenTelemetry::Instrumentation::Rack.context_with_span(span, parent_context: ctx)
+
+            contexts = [frontend_context, ctx, rack_ctx]
+            contexts.compact!
+
+            tokens = contexts.map { |context| OpenTelemetry::Context.attach(context) }
+
+            request.env[TOKENS_KEY] = tokens
+          end
+
+          # Optionally adds debugging response headers injected from {response_propagators}
+          #
+          # @param [Rack::Request] The current HTTP request
+          # @param [Rack::Response] This current HTTP response
+          # @return [void]
+          def on_commit(request, response)
+            span = OpenTelemetry::Instrumentation::Rack.current_span
+            return unless span.recording?
+
+            begin
+              response_propagators&.each { |propagator| propagator.inject(response.headers) }
```

Comment:
hmm... I should not stop here on the _first_ failure but attempt to process all of the propagators until it has gone through the entire list.